### PR TITLE
Add SITL to ArduPilot Manager

### LINKED ISSUE #98: Include SITL in ArduPilotManager
Opens the possibility for the users to test the Companion functionalities without a Pixhawk or Navigator connected.

### PR DESCRIPTION
**Pull request by rafaellehmkuhl**
Fix #98 

Use `--sitl` to activate it.

When [this PR](https://github.com/ArduPilot/ardupilot/pull/17573) gets approved we should reintroduce validation of the SITL binary.

~~Missing tests with Rasp/Navigator.~~
Tested with x86 and ARMv7 platforms.

### CODE REVIEW

File: core/services/ardupilot_manager/mavlink_proxy/MAVProxy.py [hunk 1]
```diff
@@ -37,6 +37,8 @@ def convert_endpoint(endpoint: Endpoint) -> str:
                 master_string = f"{master.place}"
             else:
                 master_string = f"{serial_endpoint_as_input(master)}"
+        if master.connection_type == EndpointType.TCPServer:
+            master_string = f"tcp:{master.place}:{master.argument}"
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/mavlink_proxy/MAVProxy.py:41]: Is there a test for it ?

File: core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def convert_endpoint(endpoint: Endpoint) -> str:
         log = f"--log {self.logdir().resolve()}"
 
         if master.connection_type == EndpointType.TCPServer:
-            return f"{self.binary()} --tcp-port {master.argument} {log} {endpoints}"
+            return f"{self.binary()} --tcp-endpoint {master.place}:{master.argument} --tcp-port 0 {log} {endpoints}"
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py:63]: Add a comment explaining about `tcp-port 0` and why is necessary

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 3]
```diff
@@ -82,6 +79,50 @@ def start_navigator(self) -> None:
     def start_serial(self, device: str) -> None:
         self.start_mavlink_manager(Endpoint("Master", self.settings.app_name, "serial", device, 115200, protected=True))
 
+    @staticmethod
+    def get_current_arch() -> str:
+        """Maps paltform.machine() outputs to docker architectures"""
+        arch_map = {"x86_64": "amd64", "aarch64": "arm", "armv7l": "arm"}  # TODO: differentiate arm64v8/ arm32v7
+        machine = platform.machine()
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:86]:
```suggestion
        local_arch = platform.machine()
```

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 4]
```diff
@@ -82,6 +79,50 @@ def start_navigator(self) -> None:
     def start_serial(self, device: str) -> None:
         self.start_mavlink_manager(Endpoint("Master", self.settings.app_name, "serial", device, 115200, protected=True))
 
+    @staticmethod
+    def get_current_arch() -> str:
+        """Maps paltform.machine() outputs to docker architectures"""
+        arch_map = {"x86_64": "amd64", "aarch64": "arm", "armv7l": "arm"}  # TODO: differentiate arm64v8/ arm32v7
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:85]: if it does not contains x86 is arm, you don't need to have all necessary cases for it, ardupilot only supports x86 and arm.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 5]
```diff
@@ -34,11 +36,25 @@ class Platform(str, Enum):
 
     Pixhawk1 = "Pixhawk1"
     Navigator = "Navigator"
+    SITL_ARM = "SITL_arm_linux_gnueabihf"
+    SITL_X86 = "SITL_x86_64_linux_gnu"
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: This is unnecessary, you can have it as:
`SITL = detect_sitl_string()`

This will avoid all logic related to both SITL versions.

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 6]
```diff
@@ -82,6 +79,50 @@ def start_navigator(self) -> None:
...
+
+    def start_sitl(self) -> None:
+        current_arch = self.get_current_arch()
+        logger.debug(f"Architecture detected: {current_arch}")
+        firmware = os.path.join(self.settings.firmware_path, "sitl_ardusub")
+        if not os.path.exists(firmware):
+            if current_arch == "amd64":
+                temporary_file = self.firmware_download.download(Vehicle.Sub, Platform.SITL_X86)
+            elif current_arch == "arm":
+                temporary_file = self.firmware_download.download(Vehicle.Sub, Platform.SITL_ARM)
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:100]: `plataform = Platform.SITL_X86 if current_arch != "arm" else platform.SITL_ARM`

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 7]
```diff
@@ -34,11 +36,25 @@ class Platform(str, Enum):
...
+
+
+class FirmwareFormat(str, Enum):
+    """Valid firmware formats to download"""
+
+    APJ = "apj"
+    ELF = "ELF"
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:47]: Is there a reason for the capitalization here ? Can you add a comment explaining ?

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 8]
```diff
@@ -34,12 +34,15 @@ def __init__(self) -> None:
...
 
-    def run(self) -> None:
+    def run(self, sitl: bool = False) -> None:
         ArduPilotManager.check_running_as_root()
 
-        while not self.start_board(BoardDetector.detect()):
-            logger.warning("Flight controller board not detected, will try again.")
-            time.sleep(2)
+        if sitl:
+            self.start_sitl()
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:41]: Is better to have two different runs than a run with a bool specific for sitl, this makes the logic a bit more confuse here.

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 4 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:85]: Just to make sure, you're saying to remove aarch64 (ARM64), correct?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: But they are different binaries.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 8 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:41]: Agree

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (patrickelectric) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: Yes, since we are not going to run X86 in ARM and ARM in x86, it does not make sense to have both, we can have the string for comparison done in runtime and remove the extra enum and all the logic behind both.

File: core/services/ardupilot_manager/mavlink_proxy/MAVProxy.py (hunk 1 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/mavlink_proxy/MAVProxy.py:41]: There is, but the tests currently only verify if the binary runs, but not it's output.

There are several limitations to the ardupilot-manager tests today. Issue #201 is tracking that and I will open a PR focused on extending and fixing them.

File: core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py (hunk 2 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py:63]: Done.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 3 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:86]: Done.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: Ok. Got it now.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 6 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:100]: Done.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 7 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:47]: I'm trying to keep the Enums 1:1 with the manifest. They are like this there.

Comment (patrickelectric) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:47]: So, can you add this information in the code documentation, otherwise this is not clear.

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:47]: Done.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 4 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:85]: Removed aarch64.

File: core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py [hunk 9]
```diff
@@ -60,7 +60,8 @@ def convert_endpoint(endpoint: Endpoint) -> str:
         log = f"--log {self.logdir().resolve()}"
 
         if master.connection_type == EndpointType.TCPServer:
-            return f"{self.binary()} --tcp-port {master.argument} {log} {endpoints}"
+            # The "--tcp-port 0" argument is used to prevent the router from opening a default tcp port on 5760
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py:63]:
```suggestion
            # The "--tcp-port 0" argument is used to prevent the router to bind by default tcp port on 5760
```

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 10]
```diff
@@ -82,6 +76,46 @@ def start_navigator(self) -> None:
     def start_serial(self, device: str) -> None:
         self.start_mavlink_manager(Endpoint("Master", self.settings.app_name, "serial", device, 115200, protected=True))
 
+    @staticmethod
+    def get_current_arch() -> str:
+        """Maps paltform.machine() outputs to docker architectures"""
+        arch_map = {"x86_64": "amd64", "armv7l": "arm"}  # TODO: differentiate arm64v8/ arm32v7
+        local_arch = platform.machine()
+        arch = arch_map.get(local_arch, None)
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:84]: Sorry if I was not clear, but what I'm saying is that:
if matches X86 is X86_64, otherwise it should be arm, the dict creates an unnecessary complex logic. 
We don't need to check each type of arm, any arm will run armv6 and since we only supports armv7, everything else that is not armv6 should run armv7.

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 11]
```diff
@@ -82,6 +76,46 @@ def start_navigator(self) -> None:
...
+        if not arch:
+            raise Exception(f"Unknown architecture ('{local_arch}') is not supported.")
+        return arch
+
+    def run_with_sitl(self) -> None:
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/ArduPilotManager.py:89]: run_with_sitl should receive an argument of vehicle type.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (patrickelectric) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: The problem behind the actual approach is that we need to have the control over the architecture inside our code, and not inside the enum for free, making the logic unnecessary inside run_with_sitl. can you provide the diff or patch for such change ?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 12]
```diff
@@ -101,11 +118,18 @@ def _validate_firmware(firmware_path: pathlib.Path) -> bool:
         Returns:
             bool: True if valid, False if otherwise.
         """
-        with open(firmware_path, "r") as firmware_file:
-            data = json.load(firmware_file)
-            keys = data.keys()
-            if "image_size" in keys and "image" in keys:
+        if firmware_path.suffix == ".apj":
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:121]: Why not `FirmwareFormat.APJ` ?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 13]
```diff
@@ -217,24 +243,34 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
...
+        logger.debug(f"Downloading following firmware: {item}")
         path = FirmwareDownload._download(item["url"])
-        if not path or not FirmwareDownload._validate_firmware(path):
-            logger.error("Unable to validate firmware file.")
-            return None
+        if firmware_format == FirmwareFormat.ELF:
+            # Make the binary executable
+            os.chmod(str(path), stat.S_IXOTH)
+
+        # TODO: Reincorporate firmware validation for SITL once Ardupilot fixes non-zero return on --help
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:270]: Add ardupilot issue number here

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:40]: Exactly, and this is how it's done. The SITL binaries are platform specific, like the firmware binaries are. Not having them on the Enums seems like creating one more logic for a problem that is exactly what the others are.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 10 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:84]: Don't agree. In your scenario, if it detects an ARM64 it will classify it as ARM and try to download the armv6/7 binary, which will not necessarily run on it.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 12 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:121]: Done.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 11 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:89]: Done.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 13 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:270]: Done.

File: core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py (hunk 9 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/mavlink_proxy/MAVLinkRouter.py:63]: Done.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 10 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:84]: In fact this whole method came from PR #62, which was approved, and I copied it here in the intent to have common logic that will enter the common library on core in some time.
I would prefer that we try to keep the same method here and elaborate on the changes in it once we  have them on the final place, if that's possible.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py]: Pushed last commit as patch. Let me know what you think.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 14]
```diff
@@ -38,8 +39,7 @@ class Platform(str, Enum):
 
     Pixhawk1 = "Pixhawk1"
     Navigator = "Navigator"
-    SITL_ARM = "SITL_arm_linux_gnueabihf"
-    SITL_X86 = "SITL_x86_64_linux_gnu"
+    SITL = "SITL"
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:42]: That is wrong, you should use a function that generates the string, check my first comment

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 14 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:42]: Ohhhh ok, now I get it. This seems more reasonable. Will change.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 5 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py]: The archtecture from last update seems better, that's true.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 14 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py]: What do you think about this last update?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 15]
```diff
@@ -20,18 +23,54 @@
     ssl._create_default_https_context = ssl._create_unverified_context
 
 
-class Vehicle(IntEnum):
+def get_current_arch() -> str:
+    """Maps paltform.machine() outputs to docker architectures"""
+    arch_map = {"x86_64": "amd64", "armv7l": "arm"}  # TODO: differentiate arm64v8/ arm32v7
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:28]: from https://github.com/bluerobotics/companion-docker/pull/289#discussion_r640801726:

They are different things, what are you using as reference is a docker PR that sadly we need to have track of the architecture to download the correct one based on docker API.

>  if it detects an ARM64 it will classify it as ARM and try to download the armv6/7 binary, which will not necessarily run on it.

That's not true, aarch64 can run both armv6 and armv7.

ArduPilot only deploys armv6 binaries, so no armv7 or aarch64, different from docker.
Also, armv6 runs in both armv7 and aarch64, so we don't need to worry.
In docker we do not support armv6 and we probably not will since some binaries are only provided in armv7.

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 16]
```diff
@@ -82,6 +75,33 @@ def start_navigator(self) -> None:
     def start_serial(self, device: str) -> None:
         self.start_mavlink_manager(Endpoint("Master", self.settings.app_name, "serial", device, 115200, protected=True))
 
+    def run_with_sitl(self, vehicle: str = "vectored") -> None:
+        firmware = os.path.join(self.settings.firmware_path, "sitl_ardusub")
```
[the remark this answers is not in the document]
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:79]: Are you talking about the firmware name?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 15 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:28]: I usually prefer the safe route, not assuming something will run if we don't know it will, but if you are confident it's not a problem, I'm ok.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:79]: If so, I don't think it makes much sense to fix the binary name if the vehicle is a argument for the method.

What do you think about `f"sitl_{vehicle}"`?

[the remark this answers is not in the document]
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:79]: Forget what I said. The sitl binary is independent of the vehicle. Let's just keep "sitl" there.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 17]
```diff
@@ -20,6 +23,12 @@
     ssl._create_default_https_context = ssl._create_unverified_context
 
 
+def get_sitl_platform() -> str:
```
Comment (Williangalvani) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:26]: maybe "get_sitl_binary_name" or something?

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 18]
```diff
@@ -82,6 +75,33 @@ def start_navigator(self) -> None:
     def start_serial(self, device: str) -> None:
         self.start_mavlink_manager(Endpoint("Master", self.settings.app_name, "serial", device, 115200, protected=True))
 
+    def run_with_sitl(self, vehicle: str = "vectored") -> None:
+        firmware = os.path.join(self.settings.firmware_path, "sitl")
+        if not os.path.exists(firmware):
+            temporary_file = self.firmware_download.download(Vehicle.Sub, Platform.SITL)
```
Comment (Williangalvani) [inline core/services/ardupilot_manager/ArduPilotManager.py:81]: we shouldn't have to download it everytime. This is less of an issue for SITL, we should have firmware available for flashing when there's no internet connectivity, as we do in the current companion.

This is scope creep, though, so we can just open an issue.

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 19]
```diff
@@ -82,6 +75,33 @@ def start_navigator(self) -> None:
...
+            assert temporary_file, "Failed to download SITL binary."
+            shutil.move(str(temporary_file), firmware)
+
+        # ArduPilot SITL binary will bind TCP port 5760 (server) and the mavlink router will connect to it as a client
+        master_endpoint = Endpoint("Master", self.settings.app_name, "tcpin", "127.0.0.1", 5760, protected=True)
```
Comment (Williangalvani) [inline core/services/ardupilot_manager/ArduPilotManager.py:86]: can we use `Endpoint.TCPServer` ?
Is that going to work right on both mavproxy and mavlinkrouter? I expected TCPClient here

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 20]
```diff
@@ -219,24 +250,34 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
...
+        logger.debug(f"Downloading following firmware: {item}")
         path = FirmwareDownload._download(item["url"])
-        if not path or not FirmwareDownload._validate_firmware(path):
-            logger.error("Unable to validate firmware file.")
-            return None
+        if firmware_format == FirmwareFormat.ELF:
+            # Make the binary executable
+            os.chmod(str(path), stat.S_IXOTH)
+
+        # TODO: Reincorporate firmware validation for SITL once Ardupilot fixes non-zero return on --help (PR #17573)
```
Comment (Williangalvani) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: That got merged to master. we could use ardusub master(dev) for sitl. I have mixed feeling about it though.
What do you think @patrickelectric ?

**Review by Williangalvani — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 18 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:81]: Didn't get it.
Currently we download the firmware only if the binary does not exist already (`if not os.path.exists(firmware)`).

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 19 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:86]: Yes. I was just replicating how it is currently, to change them all together, but I will append a commit changing that on this PR.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 20 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: But is the binary already deployed? We use the manifest to download them, so it need to be deployed to work.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 17 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:26]: Yup

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:26]: Maybe "get_sitl_platform_name"?

The platform is a object property, so I like the idea of maintaining it's name on the function.

Comment (Williangalvani) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py]: sunds good.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 20 above)
Comment (Williangalvani) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: last build was today
https://firmware.ardupilot.org/Sub/latest/SITL_x86_64_linux_gnu/

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: Perfect!
Let me try it and I update the PR.

Comment (patrickelectric) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: @Williangalvani can we cherry pick it to the ardusub version of navigator ?
If its master we are ok.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 18 above)
Comment (Williangalvani) [reply core/services/ardupilot_manager/ArduPilotManager.py:85]: I missed that somehow. my bad. the path is in the settings folder, which means it is in the host filesystem, correct? nice

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 19 above)
Comment (patrickelectric) [reply core/services/ardupilot_manager/ArduPilotManager.py]: It may be necessary to use UDP here, if we send a reboot command the TCP client/server logic may fail since it connects only once and it does not reconnect alone.
@Williangalvani you are the network guy, can you confirm this ?
It may be safer to use UDP for our master.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 21]
```diff
@@ -36,11 +45,24 @@ class Platform(str, Enum):
 
     Pixhawk1 = "Pixhawk1"
     Navigator = "Navigator"
+    SITL = get_sitl_platform_name()
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:48]: Yes, that's the way.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 20 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: Ok, so the latest version is fixed, but as it's not classified as stable yet, we are ignoring it.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 19 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py]: Ok, but is this related to sitl or are you suggesting to change it for all other masters?

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 22]
```diff
@@ -219,24 +250,34 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
...
         path = FirmwareDownload._download(item["url"])
-        if not path or not FirmwareDownload._validate_firmware(path):
-            logger.error("Unable to validate firmware file.")
-            return None
+        if firmware_format == FirmwareFormat.ELF:
+            # Make the binary executable
+            os.chmod(str(path), stat.S_IXOTH)
+
+        # TODO: Reincorporate firmware validation for SITL once Ardupilot fixes non-zero return on --help (PR #17573)
+        if platform != Platform.SITL:
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:278]: Why are you not validating SITL ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 22 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:278]: ?

**Review by patrickelectric — APPROVED** [review]
Looks good, waiting @Williangalvani to accept.
I believe that we should be validating SITL/Navigator since we are using both from master.
But I'll let @Williangalvani decide.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 22 above)
Comment (patrickelectric) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:278]: My bad, I just accepted and added my comment there.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 20 above)
Comment (Williangalvani) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:277]: Yeah. Let's use master for now. both this patch and proper navigator support are on master only.

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 19 above)
Comment (Williangalvani) [reply core/services/ardupilot_manager/ArduPilotManager.py]: Yes, depending on the implementaton on the router side, it may just give up after the connection is lost and never reconnect.
We should move to UDP if possible or at least make sure that the routers reconnect on ardusub restart.
This may affect SITL and Navigator.

Comment (patrickelectric) [reply core/services/ardupilot_manager/ArduPilotManager.py]: Just to point it, I found this problem while working around #170

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 20 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py]: Done.

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 23]
```diff
@@ -241,21 +241,26 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
...
-            if not newest_version:
-                logger.error(f"No firmware versions found for this configuration ({vehicle} and {platform}).")
-                return None
-            version = f"STABLE-{newest_version}"
-
         firmware_format = FirmwareDownload._supported_firmware_formats[platform]
 
+        # Autodetect the latest supported version.
+        # For .apj firmwares (e.g. Pixhawk), we use the latest STABLE version while for the others (e.g. SITL and
+        # Navigator) we use latest DEV.
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:248]: You are just saying what the code does but not why (the important part).

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 24]
```diff
@@ -274,10 +279,8 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
             # Make the binary executable
             os.chmod(str(path), stat.S_IXOTH)
 
-        # TODO: Reincorporate firmware validation for SITL once Ardupilot fixes non-zero return on --help (PR #17573)
-        if platform != Platform.SITL:
-            if path and not FirmwareDownload._validate_firmware(path):
-                logger.error("Unable to validate firmware file.")
-                return None
+        if path and not FirmwareDownload._validate_firmware(path):
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:282]: Can you apply this in the correct commit ?

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 18 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py:85]: No prob.
Yes, that's the default behavior. It's usually under `/root/.config/ardupilot-manager/`

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py [hunk 25]
```diff
@@ -241,21 +241,27 @@ def download(self, vehicle: Vehicle, platform: Platform, version: str = "") -> O
             logger.error(f"Specified version not found for this configuration ({vehicle} and {platform}).")
             return None
 
-        # Autodetect the latest stable version
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:204]: Commit message is wrong:
`- SITL validation is back in this commit, since the latest DEV fixes the non-zero return on --help call.`
This does not exist here.

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/firmware_download/FirmwareDownload.py (hunk 25 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/firmware_download/FirmwareDownload.py:204]: Fixed

**Review by patrickelectric — APPROVED** [review]

File: core/services/ardupilot_manager/ArduPilotManager.py (hunk 19 above)
Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py]: Apparently there's no way to output SITL to UDP, according to [the ardupilot docs](https://ardupilot.org/dev/docs/using-sitl-for-ardupilot-testing.html#sitl-with-mavproxy-udp) and the `--help`:

image

Comment (patrickelectric) [reply core/services/ardupilot_manager/ArduPilotManager.py]: try uartA as udpin

Comment (rafaellehmkuhl) [reply core/services/ardupilot_manager/ArduPilotManager.py]:
```Starting sketch 'ArduSub'
Starting SITL input
Using Irlock at port : 9005
PANIC: Invalid device path: udpin
Running: sh dumpstack.sh 89024 >dumpstack_sitl.89024.out 2>&1
Failed```

Comment (Williangalvani) [reply core/services/ardupilot_manager/ArduPilotManager.py]: while that should technically work, I'm afraid that some things could be output only to the main port.

**Review by Williangalvani — APPROVED** [review]
